File: backend/src/main/java/com/hr/employee/service/EmployeeService.java
package com.hr.employee.service;

import com.hr.employee.exception.ResourceNotFoundException;
import com.hr.employee.model.Employee;
import com.hr.employee.repository.EmployeeRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@Transactional
public class EmployeeService {
    
    @Autowired
    private EmployeeRepository employeeRepository;
    
    // Get all employees
    public List<Employee> getAllEmployees() {
        return employeeRepository.findAll();
    }
    
    // Get employee by ID
    public Employee getEmployeeById(Long id) {
        return employeeRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Employee", "id", id));
    }
    
    // Filter employees by name
    public List<Employee> getEmployeesByName(String name) {
        return employeeRepository.findByNameContainingIgnoreCase(name);
    }
    
    // Search employees by ID or name
    public List<Employee> searchEmployees(String searchTerm) {
        // Try to parse as Long for ID search
        try {
            Long id = Long.parseLong(searchTerm);
            Employee employee = employeeRepository.findById(id).orElse(null);
            if (employee != null) {
                return List.of(employee);
            }
        } catch (NumberFormatException e) {
            // Not a number, search by name
        }
        return employeeRepository.findByNameContainingIgnoreCase(searchTerm);
    }
    
    // Create new employee
    public Employee createEmployee(Employee employee) {
        return employeeRepository.save(employee);
    }
    
    // Update employee
    public Employee updateEmployee(Long id, Employee employeeDetails) {
        Employee employee = getEmployeeById(id);
        employee.setName(employeeDetails.getName());
        employee.setEmail(employeeDetails.getEmail());
        employee.setSalary(employeeDetails.getSalary());
        return employeeRepository.save(employee);
    }
    
    // Delete employee
    public void deleteEmployee(Long id) {
        Employee employee = getEmployeeById(id);
        employeeRepository.delete(employee);
    }
}

File: frontend/src/services/EmployeeService.js
// src/services/EmployeeService.js
import axios from 'axios';

const API_BASE_URL = 'http://localhost:8080/api/employees';

class EmployeeService {
    // Get all employees
    getAllEmployees() {
        return axios.get(API_BASE_URL);
    }

    // Get employee by ID
    getEmployeeById(id) {
        return axios.get(`${API_BASE_URL}/${id}`);
    }

    // Filter employees by name
    getEmployeesByName(name) {
        return axios.get(`${API_BASE_URL}?name=${name}`);
    }

    // Search employees by ID or name
    searchEmployees(searchTerm) {
        return axios.get(`${API_BASE_URL}?search=${searchTerm}`);
    }

    // Create new employee
    createEmployee(employee) {
        return axios.post(API_BASE_URL, employee);
    }

    // Update employee
    updateEmployee(id, employee) {
        return axios.put(`${API_BASE_URL}/${id}`, employee);
    }

    // Delete employee
    deleteEmployee(id) {
        return axios.delete(`${API_BASE_URL}/${id}`);
    }
}

export default new EmployeeService();